### PR TITLE
REF: move templated index.pyx code to non-template

### PR DESCRIPTION
**Pull request by jbrockmendel**
Following cleanup of libalgos, we can get a little bit of code out of the tempita file.

### CODE REVIEW

File: pandas/_libs/index.pyx [hunk 1]
```diff
@@ -681,5 +687,10 @@ cdef class BaseMultiIndexCodesEngine:
             return False
 
 
+cdef class ObjectEngine(IndexEngine):
```
Comment (jreback) [inline pandas/_libs/index.pyx]: I would move this much higher up, is it the intention NOT to inherit from this?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/index.pyx (hunk 1 above)
Comment (jreback) [reply pandas/_libs/index.pyx]: can you add a class description (ideally for other engines too, but ok for here now)

Comment (jbrockmendel) [reply pandas/_libs/index.pyx]: Yah inheritance in this file is weird;  DatetimeEngine etc all subclass Int64Engine, which isn't defined until lower in the file once the "include ..." is done.  I'll move this up and add a description, sure.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
yeah this is ok.

**Pull request merged**